### DISCUSSION TITLE
Cookie based JWT tokens

### DISCUSSION THREAD

**Discussion by joaodlf** [fastapi/fastapi#9142] [Questions]
**Description**
First of all, I want to thank you for FastAPI - It's has been a while since I have been this excited about programming for the web. FastAPI is, so far, a really interesting project.

Looking through the [documentation](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/), I can see a very clear and concise practical guide to implement JWT tokens. I can see that the access token is returned as part of the response body:

```
@app.post("/token")
async def login(form_data: OAuth2PasswordRequestForm = Depends()):
    user_dict = fake_users_db.get(form_data.username)
    if not user_dict:
        raise HTTPException(status_code=400, detail="Incorrect username or password")
    user = UserInDB(**user_dict)
    hashed_password = fake_hash_password(form_data.password)
    if not hashed_password == user.hashed_password:
        raise HTTPException(status_code=400, detail="Incorrect username or password")

    return {"access_token": user.username, "token_type": "bearer"}
```

This appears to be a requirement for the /docs to work as expected (where one can login and execute calls on the fly), this is really cool functionality, but it seems to be tied down to the response body.

**I would like to be able to set a secure and httpOnly cookie to hold the access token**, as I feel that exposing the access token as part of the response body is detrimental to the security of my application. At the same time, I would like the /docs to remain functional with a cookie based approach.

Would this be straightforward to accomplish? Is this at all supported out of the box by FastAPI?

**Comment by joaodlf** [comment]
So I actually got this to work with some simple changes, here is a small writeup for anyone that might be interested:

First thing is to update the /token route:
```
@router.post("/token", tags=["auth"])
async def auth_token(response: Response, form_data: OAuth2PasswordRequestForm = Depends()):
    
    # auth logic here...

    access_token = create_access_token(
        data={"sub": username}, expires_delta=access_token_expires
    ).decode("utf-8")

    response.set_cookie(key="access_token",value=f"Bearer {access_token}", httponly=True)

    return
```

Notice that I need to `.decode("utf-8")` the access token (else you run into `Invalid header padding` errors when decoding the jwt). I also set my httponly cookie and return an empty response.

Now we need to create a similar class to `OAuth2PasswordBearer`, this is the class FastAPI needs to find the token url:
```
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/auth/token")
```

So this is the new class we need to replace it with:
```
class OAuth2PasswordBearerWithCookie(OAuth2):
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: str = None,
        scopes: dict = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.cookies.get("access_token")

        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None

        return param
```

This is the same exact code for `OAuth2PasswordBearer` with a small modification: ` authorization: str = request.cookies.get("access_token")` - Instead of grabbing the token from the Authentication header, we get it from the `access_token` cookie.

Your code now becomes:
```
oauth2_scheme = OAuth2PasswordBearerWithCookie(tokenUrl="/auth/token")
```

This works with /docs as well since the browser always sends back httonly cookies!

**Reply by GenhaoLi** [reply]
To avoid the warnings in `OAuth2PasswordBearerWithCookie.__init__`, you can inherit from `OAuth2PasswordBearer` and skip `__init__`, and the class declaration along with the required imports would be:
```python
from typing import Optional

from fastapi import HTTPException
from fastapi.security import OAuth2PasswordBearer
from fastapi.security.utils import get_authorization_scheme_param
from starlette.requests import Request
from starlette.status import HTTP_401_UNAUTHORIZED

class OAuth2PasswordBearerWithCookie(OAuth2PasswordBearer):
    async def __call__(self, request: Request) -> Optional[str]:
		authorization: str = request.cookies.get("access_token")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param```

**Comment by wshayes** [comment]
FYI - Blog series using this approach with FastAPI by Nils de Bruin - https://medium.com/@nils_29588

**Comment by andre-dasilva** [comment]
In the future will this method be included in the docs? I think a httponly cookie is the best way to store the jwt token, and not sending in the body with json. Also thanks @joaodlf i will try your solution in my project in the next couple of days.

Edit: Tried it today and it works nicely.

**Comment by bo5o** [comment]
Isn't this approach still vulnerable to CSRF attacks?
Adding the `SameSite=Lax` flag to the cookie could mitigate most of the vulnerabilities that come with cookies (and works in most modern browsers). Otherwise you would have to add some kind of verification cookie to your frontend and implement additional logic to handle this (similar to Django csrf_token).

**Comment by andre-dasilva** [comment]
@cbows yes it is still vulnerable to CSRF. But it's XSS safe (with httponly=true), which sending in the body (and storing on the client) isn't. SameSite is not fully supported in all browsers https://caniuse.com/#search=samesite yet. But i do think as well, that it does minimize the vulnerability a lot, if you don't have some cross-origin scenarios. 

Otherwise yes you have to use a token. in fastapi you could maybe implement it in your jwt claim and store it on the client. and with every request you send it in the header and compare it with the claim

I think stuff like this would be awesome to include in the docs. fastapi and also the docs are really awesome. Thats why i think to have the best possible security in the docs is really nice and a big selling point :-)

**Comment by bo5o** [comment]
I totally agree with you. This definitely  should go in the docs. With SameSite=Lax it might even be superior to the current approach in terms of default security.

**Comment by bo5o** [comment]
According to [starlette docs](https://www.starlette.io/responses/#set-cookie), `samesite='lax'` is the default when setting cookies.

**Comment by andre-dasilva** [comment]
I have to look if it's beeing set in my cookie. But that would make it even more relevant to include this in the docs or even implement a class in fastapi

**Comment by andre-dasilva** [comment]
@cbows I just checked my cookie and it does not set samesite. current version of fastapi (0.54.1) uses starlette (0.13.2). If you check starlette the current version is 0.13.4 and there it is set in the code. so fastapi first needs to update to 0.13.4 as a dependency. right now we have to set it manually

**Comment by adamerose** [comment]
I agree the docs should be updated to an example with HttpOnly cookies since I think that's best practice to protect session tokens from XSS

**Comment by SelfhostedPro** [comment]
Just as a heads up, you could follow the way that flask_jwt_extended does things where the function that sets the tokens as cookies also creates csrf tokens:
https://flask-jwt-extended.readthedocs.io/en/stable/tokens_in_cookies/

**Comment by F1r3Hydr4nt** [comment]
FYI, @wshayes broken link (I think) should lead to this: https://archive.is/UsaXo
> I agree the docs should be updated to an example with HttpOnly cookies since I think that's best practice to protect session tokens from XSS

Perhaps @SelfhostedPro's suggestion which shows:

>  NOTE: This is just a basic example of how to enable cookies. This is
>        vulnerable to CSRF attacks, and should not be used as is. See
>        csrf_protection_with_cookies.py for a more complete example!

And leads to: https://archive.is/Bv1ub
Could be incorporated for a fully secure solution whilst maintaining old browser support

**Comment by encryptblockr** [comment]
what is latest on this?

**Comment by encryptblockr** [comment]
> FYI - Blog series using this approach with FastAPI by Nils de Bruin - https://medium.com/@nils_29588

this page is now 404

**Comment by ieferrari** [comment]
I think a better option is  to save the token in a [samtesite=Strict cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#strict)

from [owasp csrf-cheatsheet:](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#samesite-cookie-attribute)
> The Strict value will prevent the cookie from being sent by the browser to the target site in all cross-site browsing context, even when following a regular link.

afaik this will prevent csrf and xss attacks

**Comment by ghost** [comment]
Would it be an alternative to stick with a short-lived access-token in the body and add an additional long-lived refresh-token in a cookie?

**Comment by F1r3Hydr4nt** [comment]
> > FYI - Blog series using this approach with FastAPI by Nils de Bruin - https://medium.com/@nils_29588
> 
> this page is now 404

https://archive.is/UsaXo

**Reply by yanggeorge** [reply]
Thank you.

**Comment by BigSamu** [comment]
> [FEATURE REQUEST] Authentication to use HttpOnly Cookie instead of Local Storage tokens Buuntu/fastapi-react#165

Hi Joao,

Many thanks for this code. It is working perfectly with my API. However, I wonder about one important issue. Whenever you try to log out using the interface from swagger (screenshot below), the cookie still keeps in the browser letting to access the secured endpoints you have.

![image](https://user-images.githubusercontent.com/28614451/160965389-ac58099e-78cf-4c4d-8888-42218ba843e3.png)

Consider these two cases below. In the first case, you can see that the endpoint "/api/v1/users/me" should not allow access because the user is not yet authenticated (this can be checked with the lock open). However because the cookie is still in the browser, you can retrieve the data from the endpoint anyway

![image](https://user-images.githubusercontent.com/28614451/160966235-60e848f3-0e57-4a71-b407-e8004429d8d8.png)

In this second case, I delete the cookie manually from my browser. After doing that, the behaviour is the one expected (if you try to get data from the endpoint when you are not authenticated you should receive a 401 response, Unauthorized)

![image](https://user-images.githubusercontent.com/28614451/160966412-14244f1d-7068-42f5-9a12-ea0b616f90c8.png)

 **Wonder how this can be fixed in terms of whenever you logout using the Login/Logout Form you can add/destroy the cookie. Any ideas?**

**Reply by GenhaoLi** [reply]
On the Swagger UI side, it allows customization on the logout button behavior, via plugins, see https://github.com/swagger-api/swagger-ui/issues/7203. But it seems plugin can only configured with JavaScript. See [Swagger UI config doc](https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/#:~:text=for%20Swagger%20UI.-,plugins,-Unavailable) and [Swagger UI plugin API](https://swagger.io/docs/open-source-tools/swagger-ui/customization/plugin-api/)

On the FastAPI side, see [FastAPI docs on configuring Swagger UI JS settings](https://fastapi.tiangolo.com/how-to/configure-swagger-ui/?h=swagger#javascript-only-settings), but it is confusing to me how to configure them with custom JS code.

ChatGPT suggested customizing Swagger UI page by adding a separate logout button, or just override the static files, which would probably lead a big mess. See https://chatgpt.com/share/686ffd4d-4020-8012-8f46-1239fc34725a

I do believe this problem can be solved but I've given up on this

**Comment by joaodlf** [comment]
Hi @BigSamu, it's been quite a while since I've touched a project that uses fastapi, but your conclusion at the end makes sense to me - You'll need to hook into the logout endpoint (possibly via class extension as well?) and delete the cookie.

**Comment by AyushShende25** [comment]
Is there a final solution on this ??

**Comment by HandcartCactus** [comment]
> Note for Swagger UI and Swagger Editor users: Cookie authentication is currently not supported for “try it out” requests due to browser security restrictions. See [this issue](https://github.com/swagger-api/swagger-js/issues/1163) for more information. [SwaggerHub](https://swagger.io/tools/swaggerhub/) does not have this limitation.

https://swagger.io/docs/specification/v3_0/authentication/cookie-authentication/

According to the below link, the best way to make cookies work would be to authenticate via an endpoint using the cookie, and *then* navigate to the swagger UI page hosted on the same domain. You would lose any authentication UI elements in swagger, but that would work...

https://github.com/swagger-api/swagger-js/issues/1163

**Comment by phx000** [comment]
Having an `http-only` `same-site="strict"` is actually a huge pain to work with because it won't be shared across subdomains. For example, `docs.your-domain.com` won't receive cookies set to `api.your-domain.com` or `your-domain.com`.

A solid alternative is to use an access token for authentication and a refresh token for getting new access tokens. The access token lives in a tab's memory, which means it cannot be stolen using XSS. The flow is as follows:

- User logs in with email and password.
- A response is sent with a 
  - `set-cookie` header that sets a refresh token as an `http-only=True` `same-site="strict"` cookie
  -  an access token in the response's JSON body.
- The frontend saves the access token, for example, in a React Context.
- The frontend will include the token in the authorization header as `Authorization: Bearer <token>` for every request to a protected endpoint.
- When the access token expires, the browser will call `/auth/refresh` to get a new access token. This is the only endpoint in the backend that uses the refresh `http-only` cookie for authentication. If the cookie is there and correct, it will issue a new access token in a response's JSON body.

This mitigates CSRF because the browser will have to actively set the value of the Authorization header in each request, which is simply impossible to "break" with a CSRF attack because JS is needed to manipulate the request's headers to insert `Authorization: Bearer <token-from-a-variable-in-JS>`.

And it mitigates XSS because the access token is stored in memory which means other tabs can't access it, unlike local storage for example.

But keep in mind that implementing this in React will not be straight forward and for some reason there isn't a single decent tutorial out there on this.

This is an example that implements a React Context doing what was just explained.
```js
import axios from "axios";
import {createContext, useContext, useEffect, useMemo, useState} from "react";
import {useNavigate} from "react-router-dom";
import {ROUTES} from "@/routes/index.jsx";

axios.defaults.baseURL = "https://p.local/api"

const AuthContext = createContext();

/**
 * serializeToken takes the body of a successful authentication response
 * containing an access token. In my backend setup for this project, the response body also
 * includes an ISO timestamp indicating the token's expiration.
 *
 * The function returns an object that includes the token and its
 * expiration information. This allows the frontend to:
 *   - Schedule automatic refreshes before the token expires
 *   - Know if the token is stale without having to test it on a protected endpoint
 */
export function serializeToken(data) {
    const expiration = new Date(data.access_expiration)

    return {
        token: data.access,
        expiration: expiration,
        expirationPadded: new Date(expiration.getTime() - 2 * 1000) // -2 secs
    }
}

export function isTokenValid(token) {
    return token && token.expirationPadded > new Date()
}

function AuthProvider({children}) {
    const [token, setToken] = useState();
    const [error, setError] = useState();
    const [user, setUser]=useState();
    const navigate = useNavigate();

    useEffect(() => {
        refreshToken()
    }, []);

    async function refreshToken() {
        try {
            const response = await axios.post(
                "/auth/token/refresh",
                {},
                {withCredentials: true}
            )
            const newToken = serializeToken(response.data)
            setToken(newToken);
            setError(null);
            setUser(response.data.user)
            return newToken

        } catch (err) {
            if (err.response) {
                setError(err.response)
                setToken(null)
            }
            // don't put navigate here. it will also force all public routes into /login
        }
    }

    async function useFetch(method, url, data, config) {
        let tokenToUse = token

        if (!isTokenValid(token)) {
            const newAccess = await refreshToken()
            if (!newAccess) {
                navigate(ROUTES.LOGIN.absPath)
                return
            }
            tokenToUse = newAccess
        }

        if (tokenToUse) {
            const response = await axios.request({
                method,
                url,
                ...(method === "get" ? { params: data } : { data }),
                ...config,
                headers: {
                    Authorization: `Bearer ${tokenToUse.token}`,
                    ...(config?.headers || {})
                },
            })
            return response.data
        }
    }

    const contextValue = useMemo(() => ({
        token,
        setToken,
        error,
        user,
        setUser,
        refreshToken,
        useFetch,
    }), [token, error, useFetch]);

    return (
        <AuthContext.Provider value={contextValue}>{children}</AuthContext.Provider>
    );
}

export const useAuthContext = () => {
    return useContext(AuthContext);
};

export default AuthProvider;
```

This can then be used to create a ProtectedRoute component:
```js
import {Navigate, Outlet} from "react-router-dom";
import {isTokenValid, useAuthContext} from "../store/authContext.jsx";
import LoadingPage from "@/components/pages/loadingPage.jsx";
import {ROUTES} from "@/routes/index.jsx";

export default function ProtectedRoute() {
    const {token, error, refreshToken} = useAuthContext()

    if (token) {
        if (!isTokenValid(token)) {
            refreshToken()
        }
        return <Outlet/>
    }

    if (error?.status === 401 || error?.status === 403) {
        return <Navigate to={ROUTES.LOGIN.absPath}/>
    }

    return <LoadingPage/>
}
```

You can check out my full code in [this repo](https://github.com/phx000/affiliate-platform/blob/main/src/frontend/src/store/authContext.jsx). You will probably have a hard time running the project since it's unfinished, but I polished pretty well everything auth related in that frontent.
